gtd_physical_graph/src/physical_graph.c

```diff
@@ -0,0 +1,198 @@
+{
+    return graph->vertices;
+}
+
+char **get_graph_adjacency_matrix(Graph *graph)
+{
+    return graph->adjacency_matrix;
+}
+
+int get_graph_hash(Graph *graph)
+{
+    return graph->hash / 2;
+}
+
+int set_graph_maximal(Graph *graph)
+{
+    graph->is_maximal = 1;
+    return 1;
+}
+
+int get_graph_is_maximal(Graph *graph)
+{
+    return graph->is_maximal;
+}
+
+int set_edge_connected(Graph *graph, int from, int to)
+{
+    if(from < 0 || from >= graph->vertices || to < 0 || to >= graph->vertices
+        || from == to)
+    {
+        return -1;
+    }
+
+    // update hash
+    if(graph->adjacency_matrix[from][to] != CONNECTED_SYMBOL)
+    {
+        int from_degree = graph->degree[from];
+        int to_degree = graph->degree[to];
+        graph->hash += int_pow(from_degree + 1, HASH_EXP) + int_pow(to_degree + 1, HASH_EXP);
+        graph->hash -= int_pow(from_degree, HASH_EXP) + int_pow(to_degree, HASH_EXP);
+        graph->degree[from]++;
+        graph->degree[to]++;
+    }
+
+    graph->adjacency_matrix[from][to] = CONNECTED_SYMBOL;
+    graph->adjacency_matrix[to][from] = CONNECTED_SYMBOL;
+
+    return 1;
+}
+
+int set_edge_not_connected(Graph *graph, int from, int to)
+{
+    if(from < 0 || from >= graph->vertices || to < 0 || to >= graph->vertices
+        || from == to)
+    {
+        return -1;
+    }
+
+    if(graph->adjacency_matrix[from][to] == 1)
+    {
+        int from_degree = graph->degree[from];
+        int to_degree = graph->degree[to];
+        graph->hash += int_pow(from_degree - 1, HASH_EXP) + int_pow(to_degree - 1, HASH_EXP);
+        graph->hash -= int_pow(from_degree, HASH_EXP) + int_pow(to_degree, HASH_EXP);
+        if(graph->adjacency_matrix[from][to] == CONNECTED_SYMBOL)
+        {
+            graph->degree[from]--;
+            graph->degree[to]--;
+        }
+    }
+
+    graph->adjacency_matrix[from][to] = NOT_CONNECTED_SYMBOL;
+    graph->adjacency_matrix[to][from] = NOT_CONNECTED_SYMBOL;
+
+    return 1;
+}
+
+int add_vertex(Graph *graph)
+{
+    if(graph->vertices == graph->max_vertices)
+    {
+        return -1;
+    }
+
+    graph->vertices++;
+
+    return 1;
+}
+
+Graph *copy_graph(Graph *graph)
+{
+    Graph *newGraph = create_graph(graph->max_vertices, graph->vertices);
+    newGraph->hash = graph->hash;
+    int n = graph->max_vertices;
+    for(int r = 0; r < n; ++r)
+    {
+        newGraph->degree[r] = graph->degree[r];
+        for(int c = 0; c < n; ++c)
+        {
+            newGraph->adjacency_matrix[r][c] = graph->adjacency_matrix[r][c];
+        }
+    }
+    return newGraph;
+}
+
+int *get_graph_degree(Graph *graph)
+{
+    return graph->degree;
+}
+
+// ================ ALGORITHMS ================
+
+char **create_matrix(int rows, int cols)
+{
+    char **matrix = gtd_malloc(rows * sizeof(int*));
+
+    for(int r = 0; r < rows; ++r)
+    {
+        matrix[r] = gtd_malloc(cols * sizeof(char));
+    }
+
+    return matrix;
+}
+
+int int_pow(int base, int exp)
```

Comment:
Should it be in this file? Consider moving it to common or if it is static (used only in one file) make it static and remove declaration from header. There is also pow function in math.h (which we are already linking anyway)